examples/QIR/Simulation/README.md

```diff
@@ -0,0 +1,193 @@
+
+As mentioned further up, the QIR Runtime provides three interfaces to connect backends to the Runtime:
+
+- `IRuntimeDriver` : Basic runtime functions such as qubit and measurement result management.
+- `IQuantumGateSet` : The Q# instruction set. Implementation of this interface is not strictly required, as long as some instruction set is implemented, and the QIR code only calls instructions from that set (may necessitate the use of a bridge, see the [QIR Bridge](https://github.com/microsoft/qsharp-runtime/tree/main/src/Qir/Runtime#qir-bridge-and-runtime) and [previous section](#understanding-the-qir-runtime-system)).
+- `IDiagnostics` : Optional interface to provide insight into the state of a simulator or hardware backend (useful for debugging).
+
+Let's have a look at the functionality each interface provides.
+
+---
+
+`IRuntimeDriver`
+
+```cpp
+// Qubit management
+virtual Qubit AllocateQubit() = 0;
+virtual void ReleaseQubit(Qubit qubit) = 0;
+virtual std::string QubitToString(Qubit qubit) = 0;
+
+// Result management
+virtual void ReleaseResult(Result result) = 0;
+virtual bool AreEqualResults(Result r1, Result r2) = 0;
+virtual ResultValue GetResultValue(Result result) = 0;
+virtual Result UseZero() = 0;
+virtual Result UseOne() = 0;
+```
+
+---
+
+`IQuantumGateSet`
+
+```cpp
+// Elementary operations
+virtual void X(Qubit target) = 0;
+virtual void Y(Qubit target) = 0;
+virtual void Z(Qubit target) = 0;
+virtual void H(Qubit target) = 0;
+virtual void S(Qubit target) = 0;
+virtual void T(Qubit target) = 0;
+virtual void R(PauliId axis, Qubit target, double theta) = 0;
+virtual void Exp(long numTargets, PauliId paulis[], Qubit targets[], double theta) = 0;
+
+// Multi-controlled operations
+virtual void ControlledX(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledY(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledZ(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledH(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledS(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledT(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledR(long numControls, Qubit controls[], PauliId axis, Qubit target, double theta) = 0;
+virtual void ControlledExp(long numControls, Qubit controls[], long numTargets, PauliId paulis[], Qubit targets[], double theta) = 0;
+
+// Adjoint operations
+virtual void AdjointS(Qubit target) = 0;
+virtual void AdjointT(Qubit target) = 0;
+virtual void ControlledAdjointS(long numControls, Qubit controls[], Qubit target) = 0;
+virtual void ControlledAdjointT(long numControls, Qubit controls[], Qubit target) = 0;
+
+// Measurement operations
+virtual Result Measure(long numBases, PauliId bases[], long numTargets, Qubit targets[]) = 0;
+```
+
+---
+
+`IDiagnostics`
+
+```cpp
+// Internal state
+virtual void GetState(TGetStateCallback callback) = 0; // deprecated
```

Comment:
It may be worth defining `TGetStateCallback` here.

Reply:
Is it worth it given the method is deprecated in favour of the others?